File: 0720-Longest Word in Dictionary/0720-Longest Word in Dictionary.cpp
class TrieNode {
public:
    TrieNode* next[26];
    int index;
    TrieNode() {
        memset(next, NULL, sizeof(next));
        index = -1;
    }
    ~TrieNode() {
        for (TrieNode* p : next) {
            if (p) {
                delete p;
            }
        }
    }
};

class Solution {
public:
    string longestWord(vector<string>& words) {
        TrieNode* root = new TrieNode;
        for (int i = 0; i < words.size(); ++i) {
            insert(root, words[i], i);
        }
        
        string result = search(root, words);
        delete root;
        return result;
    }
    
private:
    void insert(TrieNode* root, string word, int index) {
        TrieNode* node = root;
        for (char c : word) {
            int i = c - 'a';
            if (node->next[i] == NULL) {
                node->next[i] = new TrieNode();
            }
            node = node->next[i];
        }
        node->index = index;
    }
    
    string search(TrieNode* root, vector<string>& words) {
        stack<TrieNode*> St;
        St.push(root);
        string result;
        while (!St.empty()) {
            TrieNode* node = St.top();
            St.pop();
            if (node->index != -1 || node == root) {
                if (node != root) {
                    string word = words[node->index];
                    if (word.size() > result.size() || (word.size() == result.size() && word < result)) {
                        result = word;
                    }
                }
                
                for (int i = 0; i < 26; ++i) {
                    if (node->next[i]) {
                        St.push(node->next[i]);
                    }                    
                }
            }
        }
        
        return result;
    }
};

File: 0701-0800/0720-Longest Word in Dictionary/0720-Longest Word in Dictionary.py
from functools import reduce
import collections
class Solution:
    def longestWord(self, words: 'List[str]') -> 'str':
        Trie = lambda: collections.defaultdict(Trie)
        trie = Trie()
        END = True

        for i, word in enumerate(words):
            reduce(dict.__getitem__, word, trie)[END] = i

        St = list(trie.values())
        result = ''
        while St:
            curr = St.pop()
            if END in curr:
                word = words[curr[END]]
                if len(word) > len(result) or len(word) == len(result) and word < result:
                    result = word
                St.extend([curr[letter] for letter in curr if letter != END])

        return result

class TrieNode:
    def __init__(self):
        self.next = collections.defaultdict(TrieNode)
        self.index = -1
        
class Solution2:
    def longestWord(self, words: List[str]) -> str:
        root = TrieNode()
        for i, word in enumerate(words):
            node = root
            for c in word:
                node = node.next[c]
            node.index = i
            
        St = list(root.next.values())
        result = ''
        while St:
            node = St.pop()
            if node.index != -1:
                word = words[node.index]
                if len(word) > len(result) or len(word) == len(result) and word < result:
                    result = word
                St.extend(node.next.values())
        return result
